project/ScapegoatInspectionsGenerator.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Sonar Scala Plugin
+ * Copyright (C) 2018 All contributors
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package com.mwz.sonar.scala.scapegoat
+
+import com.sksamuel.scapegoat.Inspection
+import com.sksamuel.scapegoat.Levels
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner
+import io.github.lukehutch.fastclasspathscanner.matchprocessor.SubclassMatchProcessor
+import scala.collection.mutable
+import sbt._
+import Keys._
+
+/** SBT Task that generates a managed file with all scapegoat inspections */
+object ScapegoatInspectionsGenerator {
+  val generatorTask = Def.task {
+    val log = streams.value.log
+    log.info("Generating the scapegoat inspections file")
+
+    val inspectionClass = classOf[Inspection]
+    val inspections = mutable.ListBuffer.empty[Inspection]
+
+    // We need to override the scanner class loader so it can find the scapegoat inspections
+    log.debug("[ScapegoatInspectionsGenerator] Scanning the classpath")
+    val fastCPScanner = new FastClasspathScanner(inspectionClass.getPackage.getName)
+    fastCPScanner
+      .overrideClassLoaders(inspectionClass.getClassLoader)
+      .matchSubclassesOf(
+        inspectionClass,
+        new SubclassMatchProcessor[Inspection] {
+          override def processMatch(matchingClass: Class[_ <: Inspection]): Unit = {
+            log.debug(s"[ScapegoatInspectionsGenerator] Found the inspection: ${matchingClass.getName}")
+            inspections += matchingClass.newInstance()
+          }
+        }
+      ).scan()
+
+    val AllScapegoatInspections = inspections.toList.zipWithIndex map {
+      case (inspection, idx) =>
+        s"""ScapegoatInspection(id = ${idx}, name = "${inspection.text}", description = "${inspection.explanation.getOrElse("No Explanation")}", defaultLevel = ${inspection.defaultLevel}),"""
```

Comment:
I'm wondering if we could use the class name as an id instead of the zipped index, which isn't a stable identifier and most definitely (the numbers) will change as new inspections are added/removed - class name would work better IMHO.

Reply:
Since you're using a multi-line string it would probably make sense, just for better readability of the code, to format it like in your example above.

Regarding pretty-printing, I wouldn't worry about this as long as we exclude the folder with generated sources from git. (Actually, it might be the case that scalafmt will just pick up the file and format it automatically - not the end of the world if it doesn't)

Reply:
> Regarding pretty-printing, I wouldn't worry about this as long as we exclude the folder with generated sources from git.

Indeed, it is excluded from git and from scalafmt.
Remember that the file its managed by **SBT** so it ends in a folder inside _target_.

> Since you're using a multi-line string it would probably make sense, just for better readability of the code, to format it like in your example above.

Roger.

> I'm wondering if we could use the class name as an id instead of the zipped index.

Actually yes, definitively.
I used the zipped index because I was thinking that latter we would need an Integer, but looking at the [SonarQube **API**](http://javadocs.sonarsource.org/6.7.3/apidocs/org/sonar/api/batch/rule/ActiveRules.html#findByInternalKey-java.lang.String-java.lang.String-) again I just realized that we need a String. Sorry for the confusion.
Additionally the `scapegoat-report.xml` file include the full name of the inspection (probably that's why you're proposing it), so it is perfect 👍.